### PR TITLE
Cleanup and suggestions for organization

### PR DESCRIPTION
**Pull request by 54n17y**
I saw `booru-rs` and was thinking about using it for a small `iced` app I'm making. Though for that app I need the ability to access a number of different danbooru-like sites and load images from them.

At the moment `booru-rs` looked like it provided a decent foundation for what I was looking to do. I had initially wanted to try add support for Konachan, Safebooru, and Yandere but the current way of implementing a `BooruBuilder` I found to be way to repetitive. So I make a struct, `ClientBuilder`, that is responsible for building all site clients. 

Pros:
- It will be easier to implement a generic interface for each site. (e.g. with a little more work you could have a single `ClientBuilder` that fetches data from multiple sites).
- Reduces code duplication and improves maintainability.

Cons: 
- It's messy at the moment. You need to call `.build::<Client>()` whereas before you didn't need to. (I'm working on getting rid of the need to pass the Client as a generic at the moment.)
- Breaking change. Removes `BooruBuilder`, `BooruClient`, `DanbooruClientBuilder`, ...

Aside from the major stuff I did a few small things:
- Remove `Cargo.lock` since [cargo ignores it](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries).
- Flatten out the project structure a little bit. I'm not sure having a directory for each module was necessary.
- Fix the broken Gelbooru test.
- Write a small example for the rustdoc generated docs.

@ajiiisai if you would allow, I'm interested in adding some support for other websites (and navigation between search results) after this. Also feel free to ignore this pr completely, they're just some suggestions from a random internet citizen.

### CODE REVIEW

**Comment by 54n17y** [comment]
Oh yeah, I also removed the bool parameter from `random` to make it cleaner since you'll either call the function or not.

File: src/client/mod.rs [hunk 1]
```diff
@@ -98,21 +99,17 @@ impl ClientBuilder {
 
     /// Retrieves the posts in a random order
     pub fn random(mut self) -> Self {
-        let random_tag = match self.client_type {
-            ClientType::Danbooru => "order:random",
-            ClientType::Gelbooru => "sort:random",
-        };
-        self.tags.push(random_tag.into());
+        // let random_tag = match TypeId::of::<T>() {
```
Comment (54n17y) [inline src/client/mod.rs:102]: This should be removed. It's coming in the next commit.

File: src/client/gelbooru.rs [hunk 2]
```diff
@@ -1,22 +1,26 @@
...
-impl GelbooruClient {
-    pub fn builder() -> ClientBuilder {
-        ClientBuilder::new(ClientType::Gelbooru)
-    }
+#[async_trait]
+impl Client for GelbooruClient {
+    type Post = GelbooruPost;
+
+    const URL: &'static str = "https://gelbooru.com";
+    const SORT: &'static str = "sort:random";
```
Comment (54n17y) [inline src/client/gelbooru.rs:20]: The SORT associated constant here and in the Danbooru client also need fixing. It should be just `"sort:"`

**Comment by ajiiisai** [comment]
Thanks a lot! I skimmed through the code and it looks great at first glance.
I also wanted to make a better way of making a generic client but all the ideas I had had some drawback or another. Let me take a better look a this through the weekend, but as I said at a first glance looks great.

I want to support as many sites as possible. Even some other booru like websites that are not dependant on the danbooru/gelbooru codebase, eg: Sankaku Complex etc

And again thanks a lot for the PR ^^

**Pull request merged**